src/Parsers/ASTColumnsTransformers.cpp

```diff
@@ -94,13 +95,26 @@ void ASTColumnsExceptTransformer::transform(ASTs & nodes) const
 {
     std::set<String> expected_columns;
     for (const auto & child : children)
-        expected_columns.insert(child->as<const ASTIdentifier &>().name());
+    {
+        const auto & ident = child->as<const ASTIdentifier &>();
+        String full_name = ident.name();
+        if (ident.compound())
+        {
+            if (count(full_name.begin(), full_name.end(), '.') == 2)
+                full_name = full_name.substr(full_name.find('.') + 1);
+        }
+        expected_columns.insert(full_name);
+    }
 
     for (auto it = nodes.begin(); it != nodes.end();)
     {
         if (const auto * id = it->get()->as<ASTIdentifier>())
         {
-            auto expected_column = expected_columns.find(id->shortName());
+            auto expected_column = expected_columns.find(id->name());
+
+            if (expected_column == expected_columns.end())
+                expected_column = expected_columns.find(id->shortName());
```

Comment:
So, we search by full name and if did not find, search by short name.
But how it will distinguish between `a.x` and `b.x`?

Reply:
if table `a` and `b` have the same column name `x`, and we use `select * except x from a ,b` .
it will match the first table : a.x.
Because:
if we use `select * from a, b` it will return column x and b.x.
So I think this is not a problem.